tests/cpp/test_multidevice_matmul.cpp

```diff
@@ -436,4 +436,69 @@ TEST_F(DistributedMatmulTest, AnnotateWeightOnly) {
       __FILE__);
 }
 
+// linear([M, K], [N, K]) -> [M, N]
+//
+// K, the row dimension of the weight, is sharded on DIDx. Note that LinearOp's
+// weight is of shape [column, row]. This LinearOp is decomposed into a local
+// LinearOp followed by an Allreduce.
+TEST_F(DistributedMatmulTest, RowParallelLinear) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  TensorView* x = makeContigTensor(3);
+  TensorView* w = makeContigTensor(2);
+  TensorView* y = linear(x, w);
+  fusion->addInput(x);
+  fusion->addInput(w);
+  fusion->addOutput(y);
+
+  const auto d = communicator_->size();
+  auto mesh = DeviceMesh::createForNumDevices(d);
+
+  x->setDeviceMesh(mesh);
+  x->split(-1, d, /*inner_split=*/false);
+  x->axis(-2)->parallelize(ParallelType::DIDx);
+  x->setAllocationDomain(x->getLoopDomain(), true);
+
+  w->setDeviceMesh(mesh);
+  w->split(-1, d, /*inner_split=*/false);
+  w->axis(-2)->parallelize(ParallelType::DIDx);
+  w->setAllocationDomain(w->getLoopDomain(), true);
```

Comment:
Can be combined to save some rewriting:
```suggestion
auto all_sharded_tvs = {x, w};
for (auto tv : all_sharded_tvs) {
  tv->setDeviceMesh(mesh);
  tv->split(-1, d, /*inner_split=*/false);
  tv->axis(-2)->parallelize(ParallelType::DIDx);
  tv->setAllocationDomain(x->getLoopDomain(), true);
}

Reply:
Done partially